### PR TITLE
PHOENIX-6590 Handle rollbacks in phoenix spark connector and add way to control batch wise or task wise transactions(Rajeshbabu)

### PR DESCRIPTION
**Pull request by chrajeshbabu**

### CODE REVIEW

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   8m  5s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  28m 24s |  master passed  |
| +1 :green_heart: |  compile  |   1m 42s |  master passed  |
| +1 :green_heart: |  checkstyle  |   0m 14s |  master passed  |
| +1 :green_heart: |  javadoc  |   0m 19s |  master passed  |
| +1 :green_heart: |  scaladoc  |   1m  3s |  master passed  |
| +0 :ok: |  spotbugs  |   2m 21s |  phoenix-spark-base in master has 40 extant spotbugs warnings.  |
||| _ Patch Compile Tests _ |
| -1 :x: |  mvninstall  |  11m 28s |  root in the patch failed.  |
| -1 :x: |  compile  |   0m 49s |  phoenix-spark-base in the patch failed.  |
| -1 :x: |  javac  |   0m 49s |  phoenix-spark-base in the patch failed.  |
| -1 :x: |  scalac  |   0m 49s |  phoenix-spark-base in the patch failed.  |
| -1 :x: |  checkstyle  |   0m 16s |  phoenix-spark-base: The patch generated 12 new + 151 unchanged - 2 fixed = 163 total (was 153)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  javadoc  |   0m 21s |  the patch passed  |
| +1 :green_heart: |  scaladoc  |   1m  6s |  the patch passed  |
| -1 :x: |  spotbugs  |   0m 48s |  phoenix-spark-base in the patch failed.  |
||| _ Other Tests _ |
| -1 :x: |  unit  |   1m  7s |  phoenix-spark-base in the patch failed.  |
| +1 :green_heart: |  asflicense  |   0m  9s |  The patch does not generate ASF License warnings.  |
|  |   |  58m 42s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-66/1/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix-connectors/pull/66 |
| Optional Tests | dupname asflicense scalac scaladoc unit compile javac javadoc spotbugs hbaseanti checkstyle |
| uname | Linux f0c5ac9cf35a 4.15.0-65-generic #74-Ubuntu SMP Tue Sep 17 17:06:04 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-connectors-personality.sh |
| git revision | master / d61c672 |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| mvninstall | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-66/1/artifact/yetus-general-check/output/patch-mvninstall-root.txt |
| compile | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-66/1/artifact/yetus-general-check/output/patch-compile-phoenix-spark-base.txt |
| javac | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-66/1/artifact/yetus-general-check/output/patch-compile-phoenix-spark-base.txt |
| scalac | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-66/1/artifact/yetus-general-check/output/patch-compile-phoenix-spark-base.txt |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-66/1/artifact/yetus-general-check/output/diff-checkstyle-phoenix-spark-base.txt |
| spotbugs | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-66/1/artifact/yetus-general-check/output/patch-spotbugs-phoenix-spark-base.txt |
| unit | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-66/1/artifact/yetus-general-check/output/patch-unit-phoenix-spark-base.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-66/1/testReport/ |
| Max. process+thread count | 94 (vs. ulimit of 30000) |
| modules | C: phoenix-spark-base U: phoenix-spark-base |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-66/1/console |
| versions | git=2.7.4 maven=3.3.9 spotbugs=4.1.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 46s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  28m 17s |  master passed  |
| +1 :green_heart: |  compile  |   1m 41s |  master passed  |
| +1 :green_heart: |  checkstyle  |   0m 16s |  master passed  |
| +1 :green_heart: |  javadoc  |   0m 20s |  master passed  |
| +1 :green_heart: |  scaladoc  |   1m  5s |  master passed  |
| +0 :ok: |  spotbugs  |   2m 23s |  phoenix-spark-base in master has 40 extant spotbugs warnings.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  19m  0s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 42s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 42s |  the patch passed  |
| +1 :green_heart: |  scalac  |   1m 42s |  the patch passed  |
| -1 :x: |  checkstyle  |   0m 15s |  phoenix-spark-base: The patch generated 12 new + 151 unchanged - 2 fixed = 163 total (was 153)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  javadoc  |   0m 20s |  the patch passed  |
| +1 :green_heart: |  scaladoc  |   1m  1s |  the patch passed  |
| +1 :green_heart: |  spotbugs  |   2m 34s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  |  53m 41s |  phoenix-spark-base in the patch failed.  |
| -1 :x: |  asflicense  |   0m  9s |  The patch generated 2 ASF License warnings.  |
|  |   | 115m 12s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-66/2/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix-connectors/pull/66 |
| Optional Tests | dupname asflicense scalac scaladoc unit compile javac javadoc spotbugs hbaseanti checkstyle |
| uname | Linux 55bbd1f4f422 4.15.0-65-generic #74-Ubuntu SMP Tue Sep 17 17:06:04 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-connectors-personality.sh |
| git revision | master / d61c672 |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-66/2/artifact/yetus-general-check/output/diff-checkstyle-phoenix-spark-base.txt |
| unit | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-66/2/artifact/yetus-general-check/output/patch-unit-phoenix-spark-base.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-66/2/testReport/ |
| asflicense | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-66/2/artifact/yetus-general-check/output/patch-asflicense-problems.txt |
| Max. process+thread count | 1674 (vs. ulimit of 30000) |
| modules | C: phoenix-spark-base U: phoenix-spark-base |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-66/2/console |
| versions | git=2.7.4 maven=3.3.9 spotbugs=4.1.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

File: phoenix-spark-base/src/main/java/org/apache/phoenix/spark/datasource/v2/writer/PhoenixDataWriter.java [hunk 1]
```diff
@@ -75,6 +75,7 @@
         if (tenantId != null) {
             overridingProps.put(PhoenixRuntime.TENANT_ID_ATTRIB, tenantId);
         }
+        //overridingProps.setProperty("phoenix.transactions.enabled", "true");
```
Comment (joshelser) [inline phoenix-spark-base/src/main/java/org/apache/phoenix/spark/datasource/v2/writer/PhoenixDataWriter.java:78]: nit, remove
```suggestion
```

File: phoenix-spark-base/src/it/scala/org/apache/phoenix/spark/PhoenixSparkIT.scala [hunk 2]
```diff
@@ -86,7 +86,38 @@ class PhoenixSparkIT extends AbstractPhoenixSparkIT {
 
     results.toList shouldEqual expectedResults
   }
-  
+
+  test("Can persist data into transactional tables with pheonix.transactions.enabled option") {
```
Comment (joshelser) [inline phoenix-spark-base/src/it/scala/org/apache/phoenix/spark/PhoenixSparkIT.scala:90]:
```suggestion
  test("Can persist data into transactional tables with phoenix.transactions.enabled option") {
```

**Review by joshelser — APPROVED** [review]
Two nit-picky changes. LGTM otherwise.

**Review by stoty — APPROVED** [review]
+1 LGTM, but please fix the comment typo mentioned by Josh, and add some comments te document the new behaviour to the batch size parameter.

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   6m 34s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  30m 54s |  master passed  |
| +1 :green_heart: |  compile  |   1m 47s |  master passed  |
| +1 :green_heart: |  checkstyle  |   0m 13s |  master passed  |
| +1 :green_heart: |  javadoc  |   0m 21s |  master passed  |
| +1 :green_heart: |  scaladoc  |   1m 10s |  master passed  |
| +0 :ok: |  spotbugs  |   2m 30s |  phoenix-spark-base in master has 40 extant spotbugs warnings.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  20m 24s |  the patch passed  |
| +1 :green_heart: |  compile  |   2m 14s |  the patch passed  |
| +1 :green_heart: |  javac  |   2m 14s |  the patch passed  |
| +1 :green_heart: |  scalac  |   2m 14s |  the patch passed  |
| -1 :x: |  checkstyle  |   0m 16s |  phoenix-spark-base: The patch generated 5 new + 153 unchanged - 0 fixed = 158 total (was 153)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  javadoc  |   0m 23s |  the patch passed  |
| +1 :green_heart: |  scaladoc  |   1m 19s |  the patch passed  |
| +1 :green_heart: |  spotbugs  |   3m 15s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  |  60m 31s |  phoenix-spark-base in the patch failed.  |
| -1 :x: |  asflicense  |   0m  9s |  The patch generated 2 ASF License warnings.  |
|  |   | 132m 41s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-66/4/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix-connectors/pull/66 |
| Optional Tests | dupname asflicense scalac scaladoc unit compile javac javadoc spotbugs hbaseanti checkstyle |
| uname | Linux 8f65c0eab3fb 4.15.0-147-generic #151-Ubuntu SMP Fri Jun 18 19:21:19 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-connectors-personality.sh |
| git revision | master / d61c672 |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-66/4/artifact/yetus-general-check/output/diff-checkstyle-phoenix-spark-base.txt |
| unit | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-66/4/artifact/yetus-general-check/output/patch-unit-phoenix-spark-base.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-66/4/testReport/ |
| asflicense | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-66/4/artifact/yetus-general-check/output/patch-asflicense-problems.txt |
| Max. process+thread count | 1691 (vs. ulimit of 30000) |
| modules | C: phoenix-spark-base U: phoenix-spark-base |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-66/4/console |
| versions | git=2.7.4 maven=3.3.9 spotbugs=4.1.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Pull request merged**